crates/templates/src/source.rs

```diff
@@ -0,0 +1,105 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Context};
+use tempfile::{tempdir, TempDir};
+use tokio::process::Command;
+use url::Url;
+
+use crate::directory::subdirectories;
+
+const TEMPLATE_SOURCE_DIR: &str = "templates";
+
+#[derive(Debug)]
+pub enum TemplateSource {
+    Git(Url),
+    File(PathBuf),
+}
+
+impl TemplateSource {
+    pub fn try_from_git(git_url: impl AsRef<str>) -> anyhow::Result<Self> {
+        let url_str = git_url.as_ref();
+        let url =
+            Url::parse(url_str).with_context(|| format!("Failed to parse {} as URL", url_str))?;
+        Ok(Self::Git(url))
+    }
+}
+
+pub struct LocalTemplateSource {
+    root: PathBuf,
+    _temp_dir: Option<TempDir>,
+}
+
+impl TemplateSource {
+    pub async fn get_local(&self) -> anyhow::Result<LocalTemplateSource> {
+        match self {
+            Self::Git(url) => clone_local(url).await,
+            Self::File(path) => check_local(path).await,
+        }
+    }
+
+    pub fn requires_copy(&self) -> bool {
+        match self {
+            Self::Git(_) => true,
+            Self::File(_) => false,
+        }
+    }
+}
+
+impl LocalTemplateSource {
+    pub async fn template_directories(&self) -> anyhow::Result<Vec<PathBuf>> {
+        let templates_root = self.root.join(TEMPLATE_SOURCE_DIR);
+        if templates_root.exists() {
+            subdirectories(&templates_root).with_context(|| {
+                format!(
+                    "Failed to read contents of '{}' directory",
+                    TEMPLATE_SOURCE_DIR
+                )
+            })
+        } else {
+            Err(anyhow!(
+                "Template source {} does not contain a '{}' directory",
+                self.root.display(),
+                TEMPLATE_SOURCE_DIR
+            ))
+        }
+    }
+}
+
+async fn clone_local(url: &Url) -> anyhow::Result<LocalTemplateSource> {
+    let temp_dir = tempdir()?;
+    let path = temp_dir.path().to_owned();
+
+    let url_str = url.to_string();
```

Comment:
Seems unnecessary.

Reply:
Replaced with `as_str` which saves a clone (which is what this alluded to)